core/src/test/java/org/fao/geonet/MergeUsersByUsernameDatabaseMigrationTest.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (C) 2001-2018 Food and Agriculture Organization of the
+ * United Nations (FAO-UN), United Nations World Food Programme (WFP)
+ * and United Nations Environment Programme (UNEP)
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or (at
+ * your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ *
+ * Contact: Jeroen Ticheler - FAO - Viale delle Terme di Caracalla 2,
+ * Rome - Italy. email: geonetwork@osgeo.org
+ */
+package org.fao.geonet;
+
+
+import jeeves.server.context.ServiceContext;
+import org.fao.geonet.constants.Params;
+import org.fao.geonet.domain.*;
+import org.fao.geonet.kernel.DataManager;
+import org.fao.geonet.repository.MetadataRepository;
+import org.fao.geonet.repository.specification.MetadataSpecs;
+import org.fao.geonet.repository.specification.UserGroupSpecs;
+import org.fao.geonet.utils.Xml;
+import org.jdom.Element;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.io.ByteArrayInputStream;
+import java.util.Comparator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.junit.Assert.*;
+
+public class MergeUsersByUsernameDatabaseMigrationTest extends AbstractCoreIntegrationTest {
+
+    private Map<Integer, Integer> metadataIdList;
+
+    @Before
+    public void setUpData() throws Exception {
+        metadataIdList = new LinkedHashMap<>();
+        User user1 = _userRepo.save(newUser("Username1", Profile.Guest));
```

Comment:
The test should maybe consider one user already having lower case named no ?
It looks to be failing in such case:
```

	... 124 more
Caused by: java.sql.BatchUpdateException: Batch entry 0 update Users set isenabled='y', kind=NULL, lastLoginDate='2017-01-30T11:27:03', name='Ibolya', organisation='eu_eea', profile=3, authtype='LDAP', nodeId='srv', password='', security='', surname='Bunda', username='bundaibo' where id=442124 was aborted: ERROR: duplicate key value violates unique constraint "users_username_key"
```

Reply:
It will fail if the userToKeep (based on highest profile) is not the one with the lower case name.

Reply:
Fixed proposed in https://github.com/geonetwork/core-geonetwork/pull/4516